src/pcre2_compile.c

```diff
@@ -10432,6 +10439,18 @@ if (patlen > ccontext->max_pattern_length)
   return NULL;
```

Comment:
The logic above that restrict PCRE2_LITERAL needs to be also updated for when the optimizations are disabled through the new API.

Reply:
Hmm. So it shouldn't be allowed to enable/disable optimizations when the pattern is literal? But some optimizations, like START_OPTIMIZE, still make sense on a literal pattern.

I appreciate the suggestion but don't see a good reason right now to make a change in this area.

Reply:
> So it shouldn't be allowed to enable/disable optimizations when the pattern is literal?

I think you misread the code; it is more like you can't use the PCRE2_LITERAL option unless those two optimizations are enabled. Not sure if there might be another similar restrictions but @PhilipHazel could probably explain it better.

> don't see a good reason right now to make a change in this area.

It would be safer if you do IMHO, specially considering that the "legacy" modifiers still set the "legacy" bits ONLY, so we will need to construct a new test to exercise this new codepath an therefore there is nothing in the current tests that would cover it, even if it is obviously broken, as this (and their equivalent for the other optimization) should behave the same and don't:

```
PCRE2 version 10.45-DEV 2024-06-09 (8-bit)
  re> //literal,no_auto_possess
Failed: error 192 at offset 0: invalid option bits with PCRE2_LITERAL
  re> //literal,auto_possess_off
data> 
```

Reply:
Thanks for adding further detail. I will read the relevant section of code again.

Reply:
Hmm, I looked at the code again and it still seems that my first reading was correct.

PUBLIC_COMPILE_OPTIONS lists all the bits which can normally be set in options; but if PCRE2_LITERAL is set, then only the bits in PUBLIC_LITERAL_COMPILE_OPTIONS can be set, or else compilation fails with an error. PCRE2_NO_AUTO_POSSESS and PCRE2_NO_DOTSTAR_ANCHOR are not in PUBLIC_LITERAL_COMPILE_OPTIONS, so those two options cannot be used to disable optimizations for literal patterns.

Doubtless this is because the 2 optimizations concerned do not apply to literal patterns anyways, so it doesn't make much sense to specifically disable them in that case.

Again, this raises the question: should we block users from enabling/disabling certain optimizations in the compile context for literal patterns? And again, my opinion is: no, I don't think there is a good reason to do that.

One reason is because the compile context can be re-used for multiple calls to `pcre2_compile()`. It might happen that some of those calls are for literal patterns, while others are not.

So again, from my viewpoint, this is a *desired* difference in behavior between the legacy option bits and the new API function. However, if other key contributors agree that the use of `pcre2_set_optimize()` should be restricted if the compile context is subsequently used for a literal pattern... well and good, I can implement it in that case.

Reply:
> should we block users from enabling/disabling certain optimizations in the compile context

FWIW, I didn't raise that question in any of my comments, but instead pointed out that a `pcre2_compile()` call should fail under some circumstances that are missing when using the new API.

There is no need to change `pcre2_set_optimize()` code for that to be fixed, but we must acknowledge that since the optimization flags select different codepaths inside `pcre2_compile()`, `pcre2_jit_compile()` and `pcre2_*match()`, those should be at least tested.

I also suspect that if those optimizations are safe to use when PCRE2_LITERAL is set, then the current restriction should be lifted, instead of having "different behaviour", depending on how that disable bit was set.

Reply:
> I also suspect that if those optimizations are safe to use when PCRE2_LITERAL is set, then the current restriction should be lifted

From my reading of the code, it seems obvious that those optimizations are indeed safe to use with PCRE2_LITERAL. Remember, what is not allowed with literal patterns are the legacy options which *disable* (not enable) the optimizations.

However, although this seems obvious to me, it's still better to check. Hence, I have added 3 more tests in `testinput2` which test literal patterns together with various calls to `pcre2_set_optimize`.

Reply:
> then the current restriction should be lifted

This is an interesting thought. Again, it seems fairly clear to me that the reason for the restriction is not because anything bad would happen otherwise, but simply because the options are useless when used on a literal pattern. (It could not possibly make any difference whether those optimizations were enabled or disabled on a literal pattern.) If there is more to say, perhaps @PhilipHazel might comment. In any case, I don't think this discussion needs to hold up this PR.

Reply:
> I don't think this discussion needs to hold up this PR.

Although the focus of this PR was to get the API right and that is hopefully settled, I don't think it is wise to get used to the idea of committing code that just ignores and inconsistency, specially one that was discussed and it is still "undecided".

I would expect that we take one option (Zoltan suggested that `pcre2_compile()` should error when having the options set to disable as it does now, and should not when the optimization was disabled through the new API), and make sure it is reflected on the tests and documentation and "commit" that.

You said you were sure the optimization settings were not relevant in this case, so we could instead chose the option to lift up the restriction instead.

Or we could play it safe (as I suggested originally) and add code to return an error instead.

Either way the final state should be consistent and we will hopefully get more data later and either confirm that choice or amend it as we go.

Reply:
> Although the focus of this PR was to get the API right and that is hopefully settled, I don't think it is wise to get used to the idea of committing code that just ignores and inconsistency, specially one that was discussed and it is still "undecided".

That's true. For clarity, I am not suggesting that anyone should commit anything while ignoring an inconsistency. The seeming "inconsistency" in this case has already been looked into and it's not really an inconsistency; it's an intentional difference with the behavior of the legacy options.

> I would expect that we take one option (Zoltan suggested that `pcre2_compile()` should error when having the options set to disable as it does now, and should not when the optimization was disabled through the new API), and make sure it is reflected on the tests and documentation and "commit" that.

Zoltan's suggestion is the same as the current behavior of this PR. Tests have already been added confirming that the new API does not cause an error in the case in question.

Reply:
I am a little bit lost in this conversation, but since I am mentioned, I try to summarize my preferences:

- Keep the current behavior: if a new and old option conflicts, the old option has higher priority. I.e. if an optimization is enabled in the new flags, and it is disabled in `pcre_compile`, then disable wins.
- I don't understand the error problem. The current code throws an error, if LITERAL and certain other options are set. We should keep this. The new optimization flags should not change this. I.e. if an optimization is not allowed in LITERAL, that optimization is ignored in the new option flags.

It seems the patch is doing things this way, so it is ok for me.

Reply:
> The seeming "inconsistency" in this case has already been looked

My bad, should had used a different word; the "inconsistency" I am referring to, is that the code and documentation don't match, let me know if you need me to provide a patch.

> Zoltan's suggestion is the same as the current behavior of this PR.

I know, and that is why I mentioned it as the first option that we could take.
We could also decide on a combination of them, for example letting the PCRE2_NO_DOTSTAR_ANCHOR to not fail pcre2_compile(), is also a very minimal change and apparently safe, from your look at the code.

> Tests have already been added confirming that the new API does not cause an error

A test that shows the error from `pcre2_compile()` while using the "legacy" modifiers as I proposed in the original comments would be nice as well, since we are "committing" to that as the expected behaviour going forward.
Tests that show possible edge cases reflected from your look at the code interesting as well.

Reply:
Carlo, I'm happy to update the documentation, but I'm not sure which part is currently not matching the code. Can you clarify?

I will add tests showing the error when the legacy modifiers are used.

Reply:
> The new optimization flags should not change this. I.e. if an optimization is not allowed in LITERAL, that optimization is ignored in the new option flags.

Instead of ignoring to disable the optimization, it is actually disabling it (even if it is known to conflict with LITERAL)

@alexdowad: could this be corrected and documented?

FWIW the test part of that documentation should show (maybe using `B`, as it is done for other cases) that the optimization isn't disabled when the literal modifier is used.

Reply:
Carlo, thanks for explaining further what you meant. Further tests have been added showing what happens when the legacy options are used.

I'm glad that you are carefully checking what is needed for documentation. I wholeheartedly agree that for a library like PCRE2, documentation is of utmost importance. We both feel the same on that issue. 👍🏻 

For the documentation, I can say there is nothing which needs to be added regarding this issue. I hope this does not give the (false) impression that I want to push this PR through quickly, while sweeping issues under the rug; that is not the case. I'm not sure to what extent you understand the auto-possess and dotstar anchor optimizations, but please bear with me as I explain these. I believe that if the following points are understood, we will be able to successfully conclude this discussion, with everyone on the same page.

Let's start with AUTO_POSSESS. What does this optimization do? It takes a regex like `/x+y/` and converts it to `/x++y/`. This does not help performance when the regex matches, but when it doesn't match, the amount of useless backtracking is reduced, so the match can fail faster and move on to another position.

Now, could that optimization ever apply to a literal regex? No, it is impossible! Why? Because a literal regex can never contain any quantifiers! So how would it be possible to convert quantifiers to possessive ones? Remember, a literal regex can **only** contain literal characters. The regex `/x+y/`, if compiled with PCRE2_LITERAL, will just match a literal `x`, a literal `+`, and a literal `y`.

How about DOTSTAR_ANCHOR? This optimization takes a regex like `/.*abc/` and marks it as matching only at the beginning of a line. Then the regex engine won't try matching it in the middle of a line; it will skip to the beginning of the next line instead.

Could that ever apply to a literal regex? No, it is impossible! But why? Think about it: a literal regex can never contain META_DOT (it can only contain a literal period).  And a literal regex can never include the quantifier star either (it can only contain a literal asterisk). So how could dotstar anchoring ever be done to a literal regex? It cannot!

So both AUTO_POSSESS and DOTSTAR_ANCHOR, regardless of whether they are enabled or disabled, could never have any effect whatsoever on a literal regex. They are completely just no-ops in that case.

Does it make more sense now why `pcre2_compile()` does not allow the PCRE2_NO_AUTO_POSSESS and PCRE2_NO_DOTSTAR_ANCHOR options to be used for literal patterns? It is because these options are useless on literal regexes.

Does it also make more sense now why `pcre2_set_optimize()` allows PCRE2_{AUTO_POSSESS,DOTSTAR_ANCHOR}{,_OFF} regardless of whether we are compiling literal regexes or not? It is because the `pcre2_compile_context` can be re-used multiple times, and even if it is used to compile a literal regex now, the same compile context might be used for a non-literal regex next time.

Note this point: The `options` argument to `pcre2_compile` fundamentally applies *to one compilation* only. However, the `directive` argument to `pcre2_set_optimize` applies to *a series of one or more compilations*. So it is not surprising that while certain useless combinations are restricted from appearing in `options`, `directive` is handled a bit differently.

The documentation already accurately explains that only a few options may be used together with PCRE2_LITERAL, not including PCRE2_NO_{AUTO_POSSESS,DOTSTAR_ANCHOR}. That is good! At the same time, the documentation does not say anything about a relationship between `pcre2_set_optimize()` and `PCRE2_LITERAL`, because no such relationship exists. That is correct, and intended.

Hopefully it is clearer now that `pcre2_set_optimize` is not "disabling the optimization rather than ignoring to disable it". It is also not true that these optimizations "are known to conflict with PCRE2_LITERAL".

Carlo, your code review on this PR has been great! While this particular issue was a "red herring" (meaning something which was distracting and not relevant), much of your other feedback was excellent. I can only say "2 thumbs up". 👍🏻 👍🏻

Reply:
Thanks, and indeed (again) kudos in a very good "first" contribution.

I don't think it is a "red herring" though, as we both missed the point Zoltan was making originally and that implied the code also needed an update.

To make discussion on the specifics of that easier had prepared it as alexdowad/pcre2#8, which completes the suggested behavior in the code and documents it so users of the new API aren't surprised by the "apparent" difference of behaviour between the legacy and new codepaths.